docs/design_proposals/ko-builder.md

```diff
@@ -0,0 +1,240 @@
+# ko builder
+
+* Author(s): Halvard Skogsrud (@halvards)
+* Design Shepherd: \<skaffold-core-team-member\>
+* Date: 2021-03-29
+* Status: Draft
+
+## Objectives
+
+Fast, standardized, reproducible, configuration-less, Docker-less, and
+secure-by-default container image builds for Go apps.
+
+## Background
+
+[ko](https://github.com/google/ko) is a container image builder for Go. It's
+[fast](https://cloud.google.com/blog/topics/developers-practitioners/ship-your-go-applications-faster-cloud-run-ko),
+doesn't use a `Dockerfile` or rely on the Docker daemon, and uses
+[distroless](https://github.com/GoogleContainerTools/distroless) base images by
+default. ko is to Go apps what
+[Jib](https://github.com/GoogleContainerTools/jib) is to JVM-based apps
+(approximately).
+
+The [Knative](https://knative.dev/) and [Tekton](https://tekton.dev/) open
+source projects use ko.
+
+## Proposal
+
+This proposal adds a new `ko` builder to Skaffold, based on the `ko publish`
+command. The integration does _not_ include other ko functionality related to
+[rendering](https://github.com/google/ko#ko-resolve) manifests,
+[deploying](https://github.com/google/ko#ko-apply) to Kubernetes clusters, and
+[file watching](https://github.com/google/ko/blob/f7df8106196518df5c6c35432843421e33990329/pkg/commands/resolver.go#L240).
```

Comment:
If we going to support `ko` as a native builder, it makes sense to also figure and support file watching and sync strategy.  It will only make the case stronger. 
IIUC, With `ko` file watcher, would find yamls affected when a source code is changed and only apply/redeploy the affected resources. If my understanding correct, it can help us address one of the Friction point in this issue. 
https://github.com/GoogleContainerTools/skaffold/issues/5540
Skaffold can also benefit by not restarting port forwards for unchanged pods, reinitializing pod watchers again etc for `ko`

Reply:
ko's built-in file watch support is available for `ko apply`, `ko create` and `ko resolve`, not for `ko publish`.

The intention with this proposal was to only add ko as a builder (`ko publish`). `apply` and `resolve` enters render and deploy territory, where there's overlap in ko and Skaffold functionality.

W.r.t. #5540, the reproducibility of ko helps, especially when combined with the `sha256` tag policy. This combo would result in image references staying the same for services that haven't changed - so there shouldn't be a redeployment.